hazelcast/include/hazelcast/client/spi/impl/AbstractInvocationFuture.h

```diff
@@ -0,0 +1,550 @@
+namespace hazelcast {
+    namespace client {
+        namespace spi {
+            namespace impl {
+                /**
+                 *
+                 * @param <T>
+                 */
+                template<typename T>
+                class AbstractInvocationFuture
+                        : public InternalCompletableFuture<T>,
+                          public boost::enable_shared_from_this<AbstractInvocationFuture<T> > {
+                public:
+                    virtual ~AbstractInvocationFuture() {
+                    }
+
+                    virtual void andThen(const boost::shared_ptr<ExecutionCallback<T> > &callback) {
+                        andThen(callback, defaultExecutor);
+                    }
+
+                    virtual void andThen(const boost::shared_ptr<ExecutionCallback<T> > &callback,
+                                         const boost::shared_ptr<Executor> &executor) {
+                        Preconditions::isNotNull(callback, "callback");
+                        Preconditions::isNotNull(executor, "executor");
+
+                        boost::shared_ptr<BaseState> waiter(new ExecutionCallbackState(callback));
+                        boost::shared_ptr<BaseState> response = registerWaiter(waiter, executor);
+                        if (response != VOIDOBJECT) {
+                            unblock(callback, executor);
+                        }
+                    }
+
+                    virtual bool cancel(bool mayInterruptIfRunning) {
+                        return complete(boost::shared_ptr<IException>(
+                                new concurrent::CancellationException("AbstractInvocationFuture::cancel",
+                                                                      "Task was cancelled.")));
+                    }
+
+                    virtual bool isCancelled() {
+                        const boost::shared_ptr<BaseState> currentState = state.get();
+                        return (currentState->getType() == BaseState::Exception &&
+                                boost::static_pointer_cast<ExceptionState>(
+                                        currentState)->getException()->getErrorCode() == protocol::CANCELLATION);
+                    }
+
+                    virtual bool isDone() {
+                        return isDone(state);
+                    }
+
+                    virtual boost::shared_ptr<T> get() {
+                        boost::shared_ptr<ThreadState> thread = boost::shared_ptr<ThreadState>(
+                                new ThreadState(util::getCurrentThreadId()));
+                        boost::shared_ptr<BaseState> response = registerWaiter(thread, boost::shared_ptr<Executor>());
+                        if (response != VOIDOBJECT) {
+                            // no registration was done since a value is available.
+                            return resolveAndThrowIfException(response);
+                        }
+
+                        for (;;) {
```

Comment:
what happens if the thread is interrupted here?

Reply:
There is no interruption of threads similar to Java. This is how it works: if client is being shutdown, all the awaiting futures are completed with an exception and hence it continues out of the loop.